File: 백준/Gold/2749. 피보나치 수 3/피보나치 수 3.java
import java.util.Scanner;

public class Main {
    static final int MOD = 1000000;
    static final int P = 15 * (MOD / 10);
    static int[] dp = new int[P + 4];

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        if (n <= 1) {
            System.out.println(n);
        } else {
            init();
            System.out.println(dp[(int)(n % P)]);
        }
    }

    static int f(int n) {
        if (n <= 1) return n;
        if (dp[n] != 0) return dp[n];
        return (f(n - 1) + f(n - 2)) % MOD;
    }

    static void init() {
        for (int i = 0; i <= P; i++) {
            dp[i] = f(i);
        }
    }
}

File: 백준/Gold/2749. 피보나치 수 3/피보나치 수 3.cc

#include <iostream>
using namespace std;
typedef long long ll;
ll n;
const int MOD = 1000000;
const int P = 15 * (MOD /10);
int dp[P+4];
// 주기는 15*(MOD/10)
int f(int n) {
	if (n <= 1) return n;
	if (dp[n]) return dp[n];
	return (f(n - 1) + f(n - 2))%MOD;
}
void init() {
	for (int i = 0; i <= P; i++) {
		dp[i] = f(i);
	}
}
int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	cin >> n;
	if (n <= 1) cout << n << "\n";
	else {
		init();
		cout << dp[n%P] << "\n";
	}
	return 0;
}
